src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleObject.cs

```diff
@@ -92,9 +108,70 @@ private static void HandleEndObject(ref ReadStack state)
             }
             else
             {
+                // Set directly when handling non-nested preserved array
+                bool setPropertyDirectly = state.Current.IsPreservedArray && !state.Current.IsNestedPreservedArray;
                 state.Pop();
 
-                ApplyObjectToEnumerable(value, ref state);
+                ApplyObjectToEnumerable(value, ref state, setPropertyDirectly);
+            }
+        }
+
+        private static object GetPreservedArrayValue(ref ReadStack state)
+        {
+            // Preserved JSON arrays are wrapped into JsonPreservedReference<T> where T is the original type of the enumerable
+            // and Values is the actual enumerable instance being preserved.
+            JsonPropertyInfo info = state.Current.JsonClassInfo!.PropertyCache!["Values"];
```

Comment:
This means you couldn't have an object with property name of "Values"

Reply:
isn't the cache exclusive for the `JsonClassInfo`? 

And at this point, my `JsonClassInfo.Type` is `JsonPreserveReference<T>` which is an internal type.

Are you saying that regardless of the type, the PropertyCache is used across types?

Reply:
> Are you saying that regardless of the type, the PropertyCache is used across types

The cache itself is global (because instances of ClassInfo are held in a static dictionary).

Each entry in the cache represents a Type, so if the type (a POCO) had a property names "Values" it would collide with this.

Reply:
> The cache itself is global (because instances of ClassInfo are held in a static dictionary).

The only way you can hit this path is when your `JsonClassInfo` represent the `JsonPreservedReference<T>` internal type, its PropertyCache at this point will only contain the [`Values` property](https://github.com/dotnet/runtime/pull/655/files/c965e2cb60093c32139ef6543b8be6c6068a87df#diff-7d074ddc1c16d67315f2399ce9e1c2a6R13) I don't see how that conflicts with the property cache of other types.

Perhaps I am not explaining myself nor understanding your comment correctly?

Reply:
I see - so there's a Push here. I was assuming to push and sharing of the previous object.

Reply:
Per offline discussion, if there's only one JsonPropertyInfo in that dictionary, please use the array itself (`JsonPropertyInfo[]? PropertyCacheArray`) -- it'll avoid the expensive lookup -- but also add a `Debug.Assert(PropertyCacheArray[0] == PropertyCache!["Values"]` to catch any issues in the future.